### ISSUE TITLE
CSE became less powerfull

### ISSUE THREAD

**Issue by Electricks94** [sympy/sympy#23297]
Since [this commit](https://github.com/sympy/sympy/commit/19c715920e52337d124dda63abdc056f56ec094e) I have noticed a negative change when using the CSE in SymPy. The following example should demonstrate my problem:

```python
import sympy as sp
var = sp.symbols("a b c d")

test = -var[0] + var[3]/9 + var[2]**2/2 + var[2]/3, -var[1] + var[3]/9 + var[2]**2/2 - var[2]/3
sp.cse(test)
>>> ([(x0, c/3), (x1, c**2/2 + d/9)], [-a + x0 + x1, -b - x0 + x1])
```

After 19c715920e52337d124dda63abdc056f56ec094e:

```python
import sympy as sp
var = sp.symbols("a b c d")

test = -var[0] + var[3]/9 + var[2]**2/2 + var[2]/3, -var[1] + var[3]/9 + var[2]**2/2 - var[2]/3
sp.cse(test)
>>> ([(x0, c/3), (x1, c**2)], [-a + d/9 + x0 + x1/2, -b + d/9 - x0 + x1/2])
```

The second version does not find the replacements for the divisions and is yields more complicated terms than the first old version. Thus from my perspective, the CSE has lost some of its power

**Comment by asmeurer** [comment]
CC @smichr

**Comment by smichr** [comment]
Identification of sub-expressions within Mul or Add is not something that is guaranteed by cse. Combinatorially, this is a hard problem. If you look at the cited changes you will see that the possibility of making that identification improved in other expressions while "regressing" in the case given by the OP expression.

If you want to give cse a hint, add the subexpression that you want extracted. Here I show the results of adding `+/-` the pattern:
```python
>>> test+=(-su,)
>>> cse(test)
([(x0, c/3), (x1, d/9), (x2, c**2/2), (x3, x1 + x2)], [-a + x0 + x3, -b - x0 + x1 + x2, -x3])
>>> test=test[:2]+(su,)
>>> cse(test)
([(x0, c/3), (x1, c**2), (x2, d/9 + x1/2)], [-a + x0 + x2, -b + d/9 - x0 + x1/2, x2])
>>> test=test[:2]+(-su,su)
>>> cse(test)
([(x0, c/3), (x1, d/9), (x2, c**2/2), (x3, x1 + x2)], [-a + x0 + x3, -b - x0 + x1 + x2, -x3, x3])
```

**Comment by Electricks94** [comment]
I see the point that it is really really hard to find a generally good working approach.

In my case your solution is not really satisfying since the terms are still a bit more complicated and how do I even know the subexpression beforehand? I mean that's the whole point of the CSE extracting the common subexpressions. If I need to find out common subexpressions beforehand to give hints on how to find good common subexpressions it somewhat loses its idea.

**Comment by Electricks94** [comment]
Thank you very much. This solution definitely solves my problem.

**Comment by smichr** [comment]
The code has been cleaned up a bit just now.

**Comment by smichr** [comment]
Optimizations would include only handling Adds that share 2 or more args in common.

**Comment by asmeurer** [comment]
We should try to get some more interesting test cases for cse(). The commit in question didn't affect any test cases (unless they were updated in a later commit?), only one doctest. Ideally we should be testing many large, nontrivial expressions. The current test_cse.py doesn't really have any "large" expressions, even though this is the typical use-case for cse().

**Comment by Electricks94** [comment]
I agree with this point. I use the CSE in the back of [lbmpy](https://pypi.org/project/lbmpy/). With LBM quite huge equations arise which I simplify for example with SymPy`s CSE. In lbmpy, the change of the CSE has quite some impact.

**Comment by oscarbenjamin** [comment]
Could you perhaps distil an example from lbmpy into either a test case or something that could be added to sympy's benchmark suite?

I added some examples from pydy to the benchmarks here:
https://github.com/sympy/sympy_benchmarks/blob/c9e500f8c7a2c9ed6496103dc7c1278e8a900635/benchmarks/solve.py#L405

**Comment by smichr** [comment]
Examples are nice, but simplification that happens incidentally without defining what `CSE` should do is not a good strategy for improving `cse`. Even if we accept that the optimal simplification need not happen, searching for a greedy solution to the problem of this issue is still expensive. The very modest 4 equations of my first example currently give a `cse` result orders of magnitude faster than the greedy Add-sub-expression finding algorithm that I gave. If we include Add, we should also include Mul that is non-commutative aware.

**Comment by asmeurer** [comment]
cse is inherently heuristical, but I think we can at least define what it is aiming for.  The number of operations in the terms ought to be minimized. There could be ties here, in which case perhaps certain operations should have more weight (e.g., numerically a function call will be more expensive than addition or multiplication). 

So for instance, `x*y + x*y*z` has 4 operations (one addition and three multiplications). The cse `([(x0, x*y)], [x0*z + x0])` has 3 (one multiplication in the x0 term, and one multiplication and on addition in the final term). "Variable assignment" isn't an operation that should be counted, and in general I don't think it's a good idea to try to minimize the number of variables produced, at least by default (except obviously redundant variables should be avoided).

The OP example starts with 16 operations and had 11 prior to the given commit and 14 operations after, just from counting the number of `+`, `-`, `*`, `/`, `**` in the printed output.

This highlights why we need bigger test cases. It's hard to compare how the number of operations is reduced for an expression when the total number of operations isn't very big to begin with. If algorithm 1 results in 50% as many operations and algorithm 2 results in 40% as many operations, if you start with an expression with only 5 operations, they might both result in a final expression with 2 operations. But if you start with an expression with 100 operations, you can much more clearly see the difference between 40 and 50 final operations.

I don't know if it's feasible to actually have a count_ops function that can be passed into cse to try to minimize, similar to simplify(). That could be interesting, but it might not be easy to implement. But it's at least worth noting that this is what we should aim for when considering different optimizations.

**Comment by asmeurer** [comment]
Yeah, I just made those numbers up. My point is that differences that may seem insignificant for small expressions can be significant for larger ones. But for CSE, larger expressions are the norm, so we really should be optimizing for those situations.
